File: 백트래킹/Nqueens.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
void NQ(int*,int);
int check(int*,int y);
int N,flag=0;
int main(void){
    printf("해답을 알고 싶은 N-Queens 문제의 N을 입력하세요 : ");
    scanf("%d",&N);
    int* arr=(int*)malloc(sizeof(int)*N);
    
    NQ(arr,0);
    free(arr);
    return 0;
}
void NQ(int* arr, int x){
    if(flag)return;
    if(x==N){
        flag=1;
        for(int i=0;i<N;i++){
            for(int j=0;j<N;j++){
                if(j==arr[i])printf("Q ");
                else printf("_ ");
            }
            printf("\n");
        }
        return;
    }
    for(int i=0;i<N;i++){
        arr[x]=i;
        if(check(arr,x))NQ(arr,x+1);
    }
}
int check(int* arr, int y){
    for(int i=0;i<y;i++){
        if(arr[i]==arr[y] || y-i==abs(arr[i]-arr[y]))return 0;
    }
    return 1;
}

File: 백트래킹/Nqueens.py
N=int(input("해답을 알고 싶은 N-Queens 문제의 N을 입력하세요 : "))
arr=[0]*N
flag=0
def check(arr,y):
    for i in range(y):
        if arr[i]==arr[y] or y-i==abs(arr[i]-arr[y]): return False
    return True
def NQ(arr,x):
    global N,flag
    if flag==1:return
    if x==N:
        flag=1
        for i in range(N):
            for j in range(N):
                if j==arr[i]: print("Q",end=' ')
                else: print("_",end=' ')
            print()
        return
    for i in range(N):
        arr[x]=i
        if check(arr,x)==1: NQ(arr,x+1)
NQ(arr,0)